File: iosApp/iosApp/Features/Feed/FeedViewModel.swift
//
//  FeedViewModel.swift
//  BDE
//
//  Created by Nathan Fallet on 31/01/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

class FeedViewModel: ObservableObject {
    
    @Published var isNewMenuShown = false
    @Published var isNewEventShown = false
    @Published var isSendNotificationShown = false
    @Published var events = [Event]()
    @Published var topics = [Topic]()
    @Published var cotisantConfigurations = [CotisantConfiguration]()
    @Published var ticketConfigurations = [TicketConfiguration]()
    
    func onAppear() {
        AnalyticsService.shared.log(.screenView(screenName: "feed", screenClass: "FeedView"))
        
        fetchData()
    }
    
    func fetchData() {
        Task {
            let events = try await CacheService.shared.apiService().getEvents(offset: 0, limit: 10)
            DispatchQueue.main.async {
                self.events = events
            }
        }
        Task {
            let topics = try await CacheService.shared.apiService().getTopics(offset: 0, limit: 10)
            DispatchQueue.main.async {
                self.topics = topics
            }
        }
        Task {
            let cotisantConfigurations = try await CacheService.shared.apiService().getCotisantConfigurations()
            DispatchQueue.main.async {
                self.cotisantConfigurations = cotisantConfigurations
            }
        }
        Task {
            let ticketConfigurations = try await CacheService.shared.apiService().getTicketConfigurations()
            DispatchQueue.main.async {
                self.ticketConfigurations = ticketConfigurations
            }
        }
    }
    
    func showNewMenu() {
        isNewMenuShown = true
    }
    
    func showNewEvent() {
        isNewEventShown = true
    }
    
    func showSendNotification() {
        isSendNotificationShown = true
    }
    
}

File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/feed/FeedViewModel.kt
package me.nathanfallet.bdeensisa.features.feed

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.launch
import me.nathanfallet.bdeensisa.database.DatabaseDriverFactory
import me.nathanfallet.bdeensisa.extensions.SharedCacheService
import me.nathanfallet.bdeensisa.models.CotisantConfiguration
import me.nathanfallet.bdeensisa.models.Event
import me.nathanfallet.bdeensisa.models.TicketConfiguration
import me.nathanfallet.bdeensisa.models.Topic

class FeedViewModel(application: Application): AndroidViewModel(application) {

    // Properties

    private var isNewMenuShown = MutableLiveData(false)
    private var events = MutableLiveData<List<Event>>()
    private var topics = MutableLiveData<List<Topic>>()
    private val cotisantConfigurations = MutableLiveData<List<CotisantConfiguration>>()
    private val ticketConfigurations = MutableLiveData<List<TicketConfiguration>>()

    // Getters

    fun getEvents(): LiveData<List<Event>> {
        return events
    }

    fun getTopics(): LiveData<List<Topic>> {
        return topics
    }

    fun getIsNewMenuShown(): LiveData<Boolean> {
        return isNewMenuShown
    }

    fun getCotisantConfigurations(): LiveData<List<CotisantConfiguration>> {
        return cotisantConfigurations
    }

    fun getTicketConfigurations(): LiveData<List<TicketConfiguration>> {
        return ticketConfigurations
    }

    // Setters

    fun setIsNewMenuShown(value: Boolean) {
        isNewMenuShown.value = value
    }

    // Methods

    init {
        Firebase.analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, "feed")
            param(FirebaseAnalytics.Param.SCREEN_CLASS, "FeedView")
        }

        fetchData()
    }

    fun fetchData() {
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .getEvents().let {
                        events.value = it
                    }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .getTopics().let {
                        topics.value = it
                    }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .getCotisantConfigurations().let {
                        cotisantConfigurations.value = it
                    }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .getTicketConfigurations().let {
                        ticketConfigurations.value = it
                    }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

}